Assets/Scripts/GenerateMap.cs

```diff
@@ -0,0 +1,179 @@
+using System;
+using System.Collections.Generic;
+using UnityEngine;
+using Random = UnityEngine.Random;
+
+public class GenerateMap : MonoBehaviour
+{
+    [SerializeField] private Camera mainCamera;
+    [SerializeField] private GameObject NodePrefab;
+    [SerializeField] private GameObject RootNode;
+    [SerializeField] private GameObject EndNode;
+    [SerializeField] private int rows;
+    [SerializeField] private float nodeSpacing = 1f;
+    [SerializeField] private int ymin;
+    [SerializeField] private float ymax;
+    public List<GameObject> Nodes;
+    Queue<Node> queue=new Queue<Node>();
+    private List<List<GameObject>> _levelNodes=new List<List<GameObject>>();
+
+
+    void Start()
+    {
+        
+        Nodes = new List<GameObject>();
+        Node rootNode=RootNode.GetComponent<Node>();
+        rootNode.parent=null;
+        rootNode.level=0;
+        _levelNodes.Add(new List<GameObject>());
+        _levelNodes[0].Add(RootNode);
+        rootNode.ymin = ymin;
+        rootNode.ymax = ymax;
+        EndNode.GetComponent<Node>().children=null;
+        queue.Enqueue(rootNode);
+        Nodes.Add(RootNode);
+        GenerateGrid();
+        ConnectNodes();
+    }
+
+    void GenerateGrid()
+    {
+        while (queue.Count > 0)
+        {
+            Node currentParentNode = queue.Dequeue();
+            int paths;
+            if (currentParentNode.level == rows)
+            {
+                currentParentNode.children.Add(EndNode.GetComponent<Node>());
+                continue;
+            }
+            int probability = Random.Range(0, 30);
+            if (probability <= 2)
+            {
+                paths = 2;
+            }
+            else
+            {
+                paths = 1;
+            }
```

Comment:
i think it would be better to specify upper limit of road splits by a constant (or random up to this constant), so that you can avoid this rare huge tree.

for example you could generate at start `int maxPathSplit = Random.Range(3,6)` and have incrementer, something like this

```suggestion
            int probability = Random.Range(0, 30);
            if (probability <= 2 && pathSplits < maxPathSplit)
            {
                paths = 2;
                pathSplits++;
            }
            else
            {
                paths = 1;
            }
```

probably would be even better to consider counting splits per subtree from root instead of entire tree